File: 1312-minimum-insertion-steps-to-make-a-string-palindrome/1312-minimum-insertion-steps-to-make-a-string-palindrome.cpp
class Solution {
public:
    int minInsertions(string s) {

        string text1 = s;
        reverse(s.begin(),s.end());
        string text2 = s;
         vector<vector<int>>dp(text1.size()+1,vector<int>(text2.size()+1,0));
        

        for(int i=1;i<=text1.size();i++){
            for(int j=1;j<=text2.size();j++){

                int match = 0;
                if(text1[i-1] == text2[j-1]){
                    match = 1 + dp[i-1][j-1];
                }
                int notmatch = max(dp[i-1][j],dp[i][j-1]);

                dp[i][j] = max(match, notmatch);
            }
        }

        return s.size() - dp[text1.size()][text2.size()];
    }
};

File: 1312-minimum-insertion-steps-to-make-a-string-palindrome/1312-minimum-insertion-steps-to-make-a-string-palindrome.java
class Solution {
    public int minInsertions(String s) {
        
          int[][] dp = new int[s.length()+2][s.length()+2];
        
        for(int[] a: dp){
            Arrays.fill(a,-1);
        }
        String s2 = new StringBuilder(s).reverse().toString();

        for(int i=0;i<s.length();i++){
          dp[0][i] = 0;
        }
        for(int i=0;i<s.length();i++){
          dp[i][0] = 0;
        }
        int ans = 0;
        for(int i=0;i<s.length();i++){
            for(int j=0;j<s.length();j++){
                if(s.charAt(i) == s2.charAt(j)){
                    dp[i+1][j+1] = 1 + dp[i][j];
                   
                }
                else dp[i+1][j+1] = Math.max(dp[i+1][j] , dp[i][j+1]);
            }
        }

        return (s.length() - dp[s.length()][s.length()]);
    }
}